### PR TITLE
[WIP] Refactor Meta's deps structure to an array of arrays.

### PR DESCRIPTION
**Pull request by rwjblue**
/cc @krisselden @ef4

### CODE REVIEW

**Comment by rwjblue** [comment]
@krisselden - I updated this to use an array of arrays (instead of a map of arrays).  I left it in two commits so that it was easy to compare the difference (in case you end up liking the map of arrays approach better).

The best summary of the new structure is from `writeDep` where we setup the structure:

```js
    // add the subkey to the outerArray
    keyArray.push([
      subkey,
      // array of itemKeys for the subkey
      [
        [itemKey, value]
      ]
    ]);
```

**Comment by machty** [comment]
this used to be 5.4 http://numberofpeoplewhounderstandembermetal.com/

File: packages/ember-metal/lib/meta.js [hunk 1]
```diff
@@ -176,45 +176,80 @@ Meta.prototype._findInherited = function(key, subkey) {
...
+function inheritedArrayOfArrays(name, Meta) {
   let key = memberProperty(name);
   let capitalized = capitalize(name);
 
-  Meta.prototype['write' + capitalized] = function(subkey, itemkey, value) {
-    let outerMap = this._getOrCreateOwnMap(key);
-    let innerMap = outerMap[subkey];
-    if (!innerMap) {
-      innerMap = outerMap[subkey] = new EmptyObject();
+  Meta.prototype['write' + capitalized] = function(subkey, itemKey, value) {
```
Comment (chadhietala) [inline packages/ember-metal/lib/meta.js]: Nit can you do ``write${capitalized}``

File: packages/ember-metal/lib/meta.js [hunk 2]
```diff
@@ -176,45 +176,80 @@ Meta.prototype._findInherited = function(key, subkey) {
 
 export const UNDEFINED = symbol('undefined');
 
-// Implements a member that provides a lazily created map of maps,
+// Implements a member that provides a lazily created array of arrays,
 // with inheritance at both levels.
-function inheritedMapOfMaps(name, Meta) {
+function inheritedArrayOfArrays(name, Meta) {
   let key = memberProperty(name);
   let capitalized = capitalize(name);
```
Comment (chadhietala) [inline packages/ember-metal/lib/meta.js]: Is it possible to just pass the capitalized name at the call site?

File: packages/ember-metal/lib/meta.js [hunk 3]
```diff
@@ -228,20 +263,26 @@ Meta.prototype._forEachIn = function(key, subkey, fn) {
...
+    let keyArray = pointer[key];
+    if (keyArray) {
+      for (let i = 0; i < keyArray.length; i++) {
+        if (keyArray[i][0] === subkey) {
+          let subkeyArray = keyArray[i][1];
+
+          for (let j = 0; j < subkeyArray.length; j++) {
+            let innerKey = subkeyArray[j][0];
+
+            if (!seen[innerKey]) {
```
Comment (chadhietala) [inline packages/ember-metal/lib/meta.js]: Should `seen` be a `Dict`? What is the upper bound of `seen`?

**Comment by rwjblue** [comment]
Updated #13660 to migrate to using the lodash-like object interface for this functionality.  Closing this in favor of that.

@chadhietala - I'll make sure to incorporate your comments there though...

**Pull request closed without merging**